File: julia/day02.jl
module day02
  abstract type Move end
  struct Forward <: Move; n::Int; end
  struct Down <: Move; n::Int; end

  function parse_move(input::String)::Move
    cmd, num = split(input, " ")
    num = parse(Int, num)
    cmd == "up" && return Down(-num)
    cmd == "down" && return Down(num)
    cmd == "forward" && return Forward(num)
  end

  mutable struct Submarine
    h_pos::Int
    depth::Int
    aim::Int
    Submarine() = new(0, 0, 0)
  end

  function solve_part1(state::Submarine, moves::AbstractArray{Move})::Int
    for move in moves
      if move isa Down
        state.depth += move.n
      elseif move isa Forward
        state.h_pos += move.n
      end
    end
    state.h_pos * state.depth
  end

  function solve_part2(state::Submarine, moves::AbstractArray{Move})::Int
    for move in moves
      if move isa Down
        state.aim += move.n
      elseif move isa Forward
        state.h_pos += move.n
        state.depth += move.n * state.aim
      end
    end
    state.h_pos * state.depth
  end

  function solve()::Nothing
    moves = parse_move.(eachline(".input/day02"))

    part1 = solve_part1(Submarine(), moves)
    part2 = solve_part2(Submarine(), moves)

    println("day02: $part1 $part2")
  end
end

File: kotlin/day02.kt
package aoc

private sealed class Move {
  class Down(val steps: Int) : Move()
  class Forward(val steps: Int) : Move()
}

private fun String.toMove(): Move {
  val (lhs, rhs) = split(' ')
  val num = rhs.toInt()
  return when (lhs) {
    "up" -> Move.Down(-num)
    "down" -> Move.Down(num)
    "forward" -> Move.Forward(num)
    else -> unreachable()
  }
}

private class Submarine {
  var hPos = 0
  var depth = 0
  var aim = 0

  fun part1(moves: Iterable<Move>): Int {
    for (move in moves) when (move) {
      is Move.Down -> depth += move.steps
      is Move.Forward -> hPos += move.steps
    }
    return hPos * depth
  }

  fun part2(moves: Iterable<Move>): Int {
    for (move in moves) when (move) {
      is Move.Down -> aim += move.steps
      is Move.Forward -> {
        hPos += move.steps
        depth += move.steps * aim
      }
    }
    return hPos * depth
  }
}

fun day02() {
  val input = java.io.File(".input", "day02")
  val moves = input.useLines { it.map(String::toMove).toList() }
  val part1 = Submarine().part1(moves)
  val part2 = Submarine().part2(moves)
  println("day02: $part1 $part2")
}
